### PR TITLE
Implement Proposer Signature

### LINKED ISSUE #2950: Implement VerifyProposerSignature
https://github.com/prysmaticlabs/prysm/blob/e5648f46c15765e21909868be359219790094ea9/beacon-chain/core/blocks/block_operations.go#L29-L38

### PR DESCRIPTION
**Pull request by 0xKiwi**
Resolves #2950

This PR implements verifyProposerSignature which we have been lacking for a while, the functionality was actually in function with ProcessBlockHeader but was untested.

### CODE REVIEW

File: beacon-chain/core/blocks/block_operations.go [hunk 1]
```diff
@@ -103,6 +92,7 @@ func ProcessBlockHeader(
 	beaconState *pb.BeaconState,
 	block *pb.BeaconBlock,
 	verifySignatures bool,
+	enableLogging bool,
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]: Please no more boolean flags like this

File: beacon-chain/core/blocks/block_operations.go [hunk 2]
```diff
@@ -141,26 +131,47 @@ func ProcessBlockHeader(
...
 	return beaconState, nil
 }
 
+// verifyProposerSignature uses BLS signature verification to ensure
+// the correct proposer created an incoming beacon block during state
+// transition processing.
+//
+// Spec pseudocode definition:
+//   assert bls_verify(proposer.pubkey, signing_root(block), block.signature, get_domain(state, DOMAIN_BEACON_PROPOSER))
+func verifyProposerSignature(beaconState *pb.BeaconState, block *pb.BeaconBlock, proposerIdx uint64, enableLogging bool) error {
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]: No boolean flags for logging please. Either remove the log or keep it.

File: beacon-chain/core/blocks/block_operations.go [hunk 3]
```diff
@@ -141,26 +130,41 @@ func ProcessBlockHeader(
...
 	return beaconState, nil
 }
 
+// verifyProposerSignature uses BLS signature verification to ensure
+// the correct proposer created an incoming beacon block during state
+// transition processing.
+//
+// Spec pseudocode definition:
+//   assert bls_verify(proposer.pubkey, signing_root(block), block.signature, get_domain(state, DOMAIN_BEACON_PROPOSER))
+func verifyProposerSignature(beaconState *pb.BeaconState, block *pb.BeaconBlock, proposerIdx uint64) error {
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: this seems like a very standard pattern, can we generalize it to just `verifySignature`?

we can use this in `ProcessTransfers`, `ProcessBlockHeader`, `ProcessAttestation` and more...

Comment (0xKiwi) [reply beacon-chain/core/blocks/block_operations.go]: Good idea!

**Comment by prestonvanloon** [comment]
Please resolve conflicts :(

File: beacon-chain/core/blocks/block_operations.go [hunk 4]
```diff
@@ -223,29 +241,6 @@ func ProcessRandao(
 	return beaconState, nil
 }
 
-// Verify that bls_verify(proposer.pubkey, hash_tree_root(get_current_epoch(state)),
-//   block.body.randao_reveal, domain=get_domain(state.fork, get_current_epoch(state), DOMAIN_RANDAO))
-func verifyBlockRandao(beaconState *pb.BeaconState, body *pb.BeaconBlockBody, proposerIdx uint64) error {
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]: Why is this deleted?

Comment (0xKiwi) [reply beacon-chain/core/blocks/block_operations.go]: Replaced with `verifySignature()`. Rather we have general functions than specific for each operation.

**Comment by prestonvanloon** [comment]
LGTM thanks
